### ISSUE TITLE
Class Imbalance & BBox Issues + RF-DETR Request

### ISSUE THREAD

**Issue by sctrueew** [ArgoHA/custom_d_fine#5]
Hi, thank you for the nice work. I have trained a model

- model: m
- size: 832x832
- epochs: 30
- batch_size: 12
- b_accum_steps: 1
- keep_ratio: true
- amp_enabled: true

 Images in train: 4928, val: 870, test: 0, and about 1800 background images.
2025-04-15 16:54:19.902 | INFO     | src.dl.dataset:build_dataloaders:407 - Objects count: 

**class1: 2407, class2: 758, class3: 2218, class4: 88, class5: 432, class6: 20**

Can you give me some advice on how to improve? Sometimes I notice in the training image logs that some bounding boxes are misaligned and don't properly cover the objects. I've trained with mosaic_prob: 0 and mosaic_prob: 0.8, and the model performs better when mosaic_prob is set to 0.8. Is there a way to handle the imbalance in the dataset?
```
+-----+--------+--------+-----------+--------+--------+-----------+-------+-------+-------+
|     | mAP_50 |   f1   | precision | recall |  iou   | mAP_50_95 |  TPs  |  FPs  |  FNs  |
+-----+--------+--------+-----------+--------+--------+-----------+-------+-------+-------+
| val | 0.8223 | 0.8191 |  0.8289   | 0.8094 | 0.5855 |  0.5826   | 722.0 | 149.0 | 170.0 |
+-----+--------+--------+-----------+--------+--------+-----------+-------+-------+-------+
```
I'm really impressed with the pipeline. Would it be possible to integrate RF-DETR into it?

I appreciate any help you can provide.

**Comment by ArgoHA** [comment]
Hi.
1. Just as we were discussing in [this thread](https://github.com/ArgoHA/custom_d_fine/issues/4#issuecomment-2807244848), I would suggest using less background images. Use them for testing, but during training try to have less than 10% of background images.
2. You have very little amount of objects for class6. Although we use focal loss, that should help with imbalanced data, this cace might be too harsh. Can you work on the dataset and add more samples to the class6? Can you merge it with another class?
3. I was experimenting with rf-detr couple of days ago and I found it not very impressive. While having the same accuracy, it was 2 times slower (at least on my 2 datasets). With that said, I continue looking into rf-detr and other new models and when I see that there is a better model than d-fine - I will switch the model inside this repo or integrate second model, so we can choose what to use. Final note: rf-detr might be good when you are of with training bigger models and latency is not an issue

**Comment by ArgoHA** [comment]
During inference/validation you can reduce FN with lowering the confidence threshold (from 0.5 to 0.4 for example. You can also use thresholds based on the class). That will increase FPs though.

Couple of things you can try:
- duplicate images (and annotations) were you have your minor class (as you use augmentations, they won't be absolutely the same and it will give a little more weight to the minor class)
- create synthetic images: take one with your minor object, copy it and paste to other places. It is a legit augmentation method, so should help
- leave mosaic at 0.8 or even 1.0 (should not make a ton of difference though)
- if nothing helps, maybe class specific weights to the focal loss (assigning higher weight to the minor class). This is not implemented in the repo though.

I would say second point has the best chances to work. First point is easiest to try.

**Comment by ArgoHA** [comment]
Another thing - I recommend trying keep_ratio: False, as model was pretrained with that.

>  bounding boxes aren't perfectly aligned during mosaic augmentation
do you mean the ground truth (green boxes)?

**Comment by ArgoHA** [comment]
lr should not be an issue in your case. I use scheduler, so as long as you don't see blown gradients (NaNs) lrs should be fine. You can try higher of coarse if you want.

Please share some examples with bbox issue, that might be a bug to fix in the next update

**Comment by ArgoHA** [comment]
I can confirm the issue. It happens when mosaic_augs.degrees > 0. Will work on that this week, for now just have 0.0 degrees
